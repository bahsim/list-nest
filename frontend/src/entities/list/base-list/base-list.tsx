import * as React from 'react';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import { GroupHeader } from '@/shared/ui/list/group-header';

/**
 * Props for BaseList.
 */
export interface BaseListProps<T> {
  /**
   * Groupped items.
   */
  grouppedItems: {
    label: string;
    items: T[];
    rightContent?: React.ReactNode;
  }[];
  /**
   * Render function for each item.
   */
  renderItem: (item: T, groupLabel: string) => React.ReactNode;
}

/**
 * A generic, dumb list for use in features. Handles layout, no business logic.
 */
export function BaseList<T extends { id?: string }>({
  grouppedItems,
  renderItem,
}: BaseListProps<T>) {
  return (
    <>
      {grouppedItems.map((group) => (
        <React.Fragment key={group.label}>
          {group.items.length > 0 && (
            <React.Fragment key={group.label}>
              <GroupHeader label={group.label} rightContent={group.rightContent} />
              <List sx={{ width: '100%', maxWidth: 600, px: 1 }} disablePadding>
                {group.items.map((item, i) => (
                  <ListItem
                    key={item.id ?? i}
                    disableGutters
                    disablePadding
                    sx={{ display: 'block' }}
                  >
                    {renderItem(item, group.label)}
                  </ListItem>
                ))}
              </List>
            </React.Fragment>
          )}
        </React.Fragment>
      ))}
    </>
  );
}
