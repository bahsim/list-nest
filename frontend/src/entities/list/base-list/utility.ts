import type { ExpandedItem } from '@/entities/list/types/expanded-item';
/**
 * Checks if the given item (or itemId) is expanded for the specified group.
 */

interface IsItemExpandedProps {
  expandedItem: ExpandedItem | null;
  group?: string;
  itemId: string;
}

export function isItemExpanded({
  expandedItem,
  group = 'non-group',
  itemId,
}: IsItemExpandedProps): boolean {
  return !!expandedItem && expandedItem.group === group && expandedItem.id === itemId;
}
