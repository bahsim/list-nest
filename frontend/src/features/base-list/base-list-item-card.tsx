import * as React from 'react';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Checkbox from '@mui/material/Checkbox';
import Typography from '@mui/material/Typography';
import { useTheme, type Theme, alpha } from '@mui/material/styles';
import Box from '@mui/material/Box';
import type { SxProps } from '@mui/material/styles';
import { useSwipeable } from 'react-swipeable';

/**
 * Usage Example:
 * <BaseListItemCard
 *   title="Milk"
 *   checked={true}
 *   highlighted={false}
 *   onToggle={() => {}}
 *   onClick={() => {}}
 *   onSwipeLeft={() => {}}
 *   onSwipeRight={() => {}}
 *   getSwipeVisuals={({ direction, theme }) => ({ icon: MyIcon, background: theme.palette.error.main })}
 * />
 */

/**
 * Props for BaseListItemCard.
 */
export interface BaseListItemCardProps {
  /**
   * Main title or label.
   */
  readonly title: string;
  /**
   * Optional secondary text.
   */
  readonly secondaryText?: string;
  /**
   * Whether the item is checked (e.g., bought).
   */
  readonly checked?: boolean;
  /**
   * Whether the item is highlighted (e.g., current).
   */
  readonly highlighted?: boolean;
  /**
   * Whether the item is disabled (e.g., deleted).
   */
  readonly disabled?: boolean;
  /**
   * Handler for checkbox toggle.
   */
  readonly onToggle?: () => void;
  /**
   * Handler for main click/edit.
   */
  readonly onClick?: () => void;
  /**
   * Handler for swipe left.
   */
  readonly onSwipeLeft?: () => void;
  /**
   * Handler for swipe right.
   */
  readonly onSwipeRight?: () => void;
  /**
   * Function to get swipe visuals (icon component, background) based on swipe state.
   */
  readonly getSwipeVisuals?: (args: { direction: 'left' | 'right'; theme: Theme }) => {
    icon: React.ElementType;
    background: string;
  };
  /**
   * Optional style overrides.
   */
  readonly sx?: SxProps<Theme>;
}

/**
 * Minimum swipe distance in pixels to trigger swipe action.
 */
export const SWIPE_ACTION_THRESHOLD = 80;
/**
 * Alpha value for highlighted background (0.13 ~ 22 hex).
 */
const HIGHLIGHT_ALPHA = 0.13;

/**
 * Custom hook to encapsulate swipe logic for list item cards.
 * @param onSwipeLeft - Handler for swipe left action.
 * @param onSwipeRight - Handler for swipe right action.
 * @returns { translateX, swipeHandlers, absX }
 */
function useListItemSwipe({
  onSwipeLeft,
  onSwipeRight,
}: {
  onSwipeLeft?: () => void;
  onSwipeRight?: () => void;
}) {
  const [translateX, setTranslateX] = React.useState(0);
  const swipeHandlers = useSwipeable({
    onSwiping: (e) => {
      if (Math.abs(e.deltaX) > Math.abs(e.deltaY)) {
        setTranslateX(e.deltaX);
      }
    },
    onSwiped: () => {
      setTranslateX(0);
    },
    onSwipedLeft: (e) => {
      if (e.absX >= SWIPE_ACTION_THRESHOLD && onSwipeLeft) {
        onSwipeLeft();
      }
    },
    onSwipedRight: (e) => {
      if (e.absX >= SWIPE_ACTION_THRESHOLD && onSwipeRight) {
        onSwipeRight();
      }
    },
    trackMouse: true,
    preventScrollOnSwipe: false,
  });
  const absX = Math.abs(translateX);
  return { translateX, swipeHandlers, absX };
}

/**
 * A generic, dumb list item card for use in lists. Handles layout, calls handlers, no business logic.
 */
export const BaseListItemCard: React.FC<BaseListItemCardProps> = React.memo(
  ({
    title,
    secondaryText,
    checked = false,
    highlighted = false,
    disabled = false,
    onToggle,
    onClick,
    onSwipeLeft,
    onSwipeRight,
    getSwipeVisuals,
    sx,
  }) => {
    const theme = useTheme();
    const handleToggle = React.useCallback(() => {
      if (onToggle) {
        onToggle();
      }
    }, [onToggle]);
    const handleClick = React.useCallback(() => {
      if (onClick) {
        onClick();
      }
    }, [onClick]);
    const { translateX, swipeHandlers, absX } = useListItemSwipe({ onSwipeLeft, onSwipeRight });
    let actionIcon: React.ReactNode = null;
    let actionBg = 'transparent';

    if (translateX !== 0 && getSwipeVisuals) {
      const direction = translateX < 0 ? 'left' : 'right';
      const visuals = getSwipeVisuals({ direction, theme });
      const IconComponent = visuals.icon;
      const opacity = Math.min(absX / 80, 1);
      actionIcon = (
        <IconComponent sx={{ fontSize: 32, color: '#fff', opacity, transition: 'opacity 0.1s' }} />
      );
      actionBg = absX > 10 ? visuals.background : 'transparent';
    }

    return (
      <Box sx={{ position: 'relative', mb: theme.spacing(2), ...sx }} {...swipeHandlers}>
        {/* Action background and icon */}
        <Box
          sx={{
            position: 'absolute',
            inset: 0,
            zIndex: 0,
            display: 'flex',
            alignItems: 'center',
            justifyContent: translateX < 0 ? 'flex-end' : 'flex-start',
            px: 3,
            background: actionBg,
            borderRadius: 1.5,
            transition: 'background 0.15s',
            pointerEvents: 'none',
          }}
        >
          {actionIcon}
        </Box>
        <Card
          variant="outlined"
          role="checkbox"
          aria-checked={checked}
          aria-disabled={disabled}
          sx={{
            display: 'flex',
            alignItems: 'center',
            background: highlighted
              ? alpha(theme.palette.info.main, HIGHLIGHT_ALPHA)
              : 'transparent',
            boxShadow: highlighted ? theme.shadows[3] : theme.shadows[1],
            borderColor: highlighted ? theme.palette.primary.main : theme.palette.divider,
            opacity: disabled ? 0.6 : 1,
            position: 'relative',
            cursor: onClick ? 'pointer' : 'default',
            transform: `translateX(${translateX}px)`,
            zIndex: 1,
            transition:
              translateX === 0
                ? 'box-shadow 0.2s, background 0.2s, opacity 0.2s, border-color 0.2s, transform 0.2s'
                : 'none',
          }}
          onClick={handleClick}
        >
          <Checkbox
            checked={checked}
            onChange={handleToggle}
            disabled={disabled}
            color="primary"
            sx={{ ml: theme.spacing(1), zIndex: 2 }}
          />
          <CardContent
            sx={{
              flex: 1,
              display: 'flex',
              alignItems: 'center',
              p: `${theme.spacing(1)} !important`,
              pl: '0 !important',
            }}
          >
            <Typography
              variant="body1"
              sx={{
                textDecoration: disabled ? 'line-through' : 'none',
                fontWeight: highlighted ? 700 : 400,
                color: disabled ? theme.palette.text.secondary : theme.palette.text.primary,
              }}
            >
              {title}
            </Typography>
            {secondaryText && (
              <Typography variant="body2" color="text.secondary" sx={{ ml: 2 }}>
                {secondaryText}
              </Typography>
            )}
          </CardContent>
        </Card>
      </Box>
    );
  },
);
