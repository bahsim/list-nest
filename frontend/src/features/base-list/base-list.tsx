import * as React from 'react';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import type { SxProps, Theme } from '@mui/material/styles';

/**
 * Props for BaseList.
 */
export interface BaseListProps<T> {
  items: T[];
  /**
   * Render function for each item.
   */
  renderItem: (item: T, index: number) => React.ReactNode;
  /**
   * Optional style overrides.
   */
  sx?: SxProps<Theme>;
}

/**
 * A generic, dumb list for use in features. Handles layout, no business logic.
 */
export function BaseList<T>({ items, renderItem, sx }: BaseListProps<T>) {
  return (
    <List sx={sx} disablePadding>
      {items.map((item, i) => (
        <ListItem key={i} disableGutters disablePadding sx={{ display: 'block' }}>
          {renderItem(item, i)}
        </ListItem>
      ))}
    </List>
  );
} 