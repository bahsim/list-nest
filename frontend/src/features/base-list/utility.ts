import type { MainListItem } from '@ui-kit/components/types';

/**
 * Checks if the given item (or itemId) is expanded for the specified group.
 */
export function isItemExpanded(
  expandedItem: { group: string; itemId: string } | null,
  group: string,
  itemOrId: MainListItem | string
): boolean {
  const itemId = typeof itemOrId === 'string' ? itemOrId : itemOrId.id;
  return !!expandedItem && expandedItem.group === group && expandedItem.itemId === itemId;
} 