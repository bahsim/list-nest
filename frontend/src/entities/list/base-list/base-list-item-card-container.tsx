import * as React from 'react';
import type { ListItem } from '@/entities/list/types';
import type { ExpandedItem } from '@/entities/list/types/expanded-item';
import { BaseListItemCard } from '@/widgets/base-list-item-card';
import { isItemExpanded } from './utility';

export interface BaseListItemCardContainerProps {
  item: ListItem;
  group?: string;
  expandedItem: ExpandedItem | null;
  onExpand: (group: string, id: string) => void;
  checked: boolean;
  highlighted: boolean;
  canceled: boolean;
  completed: boolean;
  secondaryText: string;
  renderExpandedContent: React.ReactNode;
  onToggle?: () => void;
  onSwipeLeft?: () => void;
  onSwipeRight?: () => void;
  getSwipeVisuals?: (args: any) => any;
}

export const BaseListItemCardContainer: React.FC<BaseListItemCardContainerProps> = ({
  item,
  group = 'non-group',
  expandedItem,
  onExpand,
  checked,
  highlighted,
  canceled,
  completed,
  secondaryText,
  renderExpandedContent,
  onToggle,
  onSwipeLeft,
  onSwipeRight,
  getSwipeVisuals,
}) => {
  const isExpanded = isItemExpanded({ expandedItem, group, itemId: item.id });
  const handleExpand = () => onExpand(group, item.id);

  return (
    <BaseListItemCard
      title={item.name}
      secondaryText={secondaryText}
      checked={checked}
      highlighted={highlighted}
      canceled={canceled}
      completed={completed}
      onToggle={onToggle}
      isExpanded={isExpanded}
      onExpand={handleExpand}
      onSwipeLeft={onSwipeLeft}
      onSwipeRight={onSwipeRight}
      getSwipeVisuals={getSwipeVisuals}
      renderExpandedContent={renderExpandedContent}
    />
  );
}; 