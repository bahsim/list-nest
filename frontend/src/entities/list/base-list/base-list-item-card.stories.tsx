import React from 'react';
import { BaseListItemCard } from '@/widgets/base-list-item-card';
import { Box } from '@mui/material';
import InfoIcon from '@mui/icons-material/Info';

export default {
  title: 'features/base-list/base-list-item-card',
  component: BaseListItemCard,
};

const getSwipeVisuals = ({ direction, theme }: { direction: 'left' | 'right'; theme: any }) => ({
  icon: InfoIcon,
  background: direction === 'left' ? theme.palette.error.main : theme.palette.success.main,
});

export const Default = () => (
  <Box width={400}>
    <BaseListItemCard
      title="Milk"
      secondaryText="2L"
      checked={false}
      highlighted={false}
      onToggle={() => alert('Toggled')}
      onClick={() => alert('Clicked')}
      getSwipeVisuals={getSwipeVisuals}
      isExpanded={false}
      onExpand={() => alert('Expanded')}
    />
  </Box>
);

export const Checked = () => (
  <Box width={400}>
    <BaseListItemCard
      title="Eggs"
      secondaryText="12"
      checked
      onToggle={() => alert('Toggled')}
      getSwipeVisuals={getSwipeVisuals}
      isExpanded={false}
      onExpand={() => alert('Expanded')}
    />
  </Box>
);

export const Highlighted = () => (
  <Box width={400}>
    <BaseListItemCard
      title="Bread"
      secondaryText="12"
      highlighted
      onClick={() => alert('Clicked')}
      getSwipeVisuals={getSwipeVisuals}
      isExpanded={false}
      onExpand={() => alert('Expanded')}
    />
  </Box>
);

export const Disabled = () => (
  <Box width={400}>
    <BaseListItemCard
      title="Butter"
      secondaryText="12"
      canceled
      onToggle={() => alert('Toggled')}
      getSwipeVisuals={getSwipeVisuals}
      isExpanded={false}
      onExpand={() => alert('Expanded')}
    />
  </Box>
); 