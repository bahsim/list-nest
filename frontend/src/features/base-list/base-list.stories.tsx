import React from 'react';
import { BaseList } from './base-list';
import { BaseListItemCard } from './base-list-item-card';
import { Box } from '@mui/material';
import InfoIcon from '@mui/icons-material/Info';
import CheckCircleIcon from '@mui/icons-material/CheckCircle';
import ErrorIcon from '@mui/icons-material/Error';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import ArrowForwardIcon from '@mui/icons-material/ArrowForward';

export default {
  title: 'features/base-list/base-list',
  component: BaseList,
};

export const StringList = () => (
  <Box width={400}>
    <BaseList
      items={['Milk', 'Eggs', 'Bread']}
      renderItem={(item) => <div>{item}</div>}
    />
  </Box>
);

export const ObjectList = () => (
  <Box width={400}>
    <BaseList
      items={[{ name: 'Milk', checked: true }, { name: 'Eggs', checked: false }]}
      renderItem={(item) => <div>{item.name} {item.checked ? '✔️' : ''}</div>}
    />
  </Box>
);

const leftIcons = [InfoIcon, ErrorIcon, ArrowBackIcon];
const rightIcons = [CheckCircleIcon, ArrowForwardIcon, InfoIcon];
const leftColors = ['#1976d2', '#d32f2f', '#ffa000'];
const rightColors = ['#2e7d32', '#0288d1', '#fbc02d'];

export const WithBaseListItemCard = () => (
  <Box width={400}>
    <BaseList
      items={['Milk', 'Eggs', 'Bread']}
      renderItem={(item, i) => {
        return (
          <BaseListItemCard
            title={item}
            onToggle={() => alert(`Toggled ${item}`)}
            getSwipeVisuals={({ direction, theme }) => {
              if (direction === 'left') {
                return {
                  icon: leftIcons[i % leftIcons.length],
                  background: leftColors[i % leftColors.length],
                };
              }
              return {
                icon: rightIcons[i % rightIcons.length],
                background: rightColors[i % rightColors.length],
              };
            }}
            isExpanded={false}
            onExpand={() => alert('Expanded')}
          />
        );
      }}
    />
  </Box>
);

export const EmptyList = () => (
  <Box width={400}>
    <BaseList items={[]} renderItem={() => <div>Should not render</div>} />
  </Box>
); 