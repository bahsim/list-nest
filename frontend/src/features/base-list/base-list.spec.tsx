import React from 'react';
import { render } from '@testing-library/react';
import { BaseList } from './base-list';
import '@testing-library/jest-dom';

describe('BaseList', () => {
  it('renders items using renderItem', () => {
    const items = ['a', 'b', 'c'];
    const { getByText } = render(
      <BaseList items={items} renderItem={(item) => <span>{item}</span>} />
    );
    items.forEach((item) => {
      expect(getByText(item)).not.toBeNull();
    });
  });

  it('applies sx prop', () => {
    const items = ['x'];
    const { container } = render(
      <BaseList items={items} renderItem={(item) => <span>{item}</span>} sx={{ background: 'red' }} />
    );
    const list = container.querySelector('ul');
    expect(list).not.toBeNull();
    // Check that the MUI class is present
    expect(list!.className).toMatch(/MuiList-root/);
    // Optionally, snapshot the HTML
    expect(list).toMatchSnapshot();
  });

  it('renders nothing for empty items', () => {
    const { container } = render(
      <BaseList items={[]} renderItem={() => <span>should not render</span>} />
    );
    // Only the List element should be present
    expect(container.querySelectorAll('li').length).toBe(0);
  });
}); 