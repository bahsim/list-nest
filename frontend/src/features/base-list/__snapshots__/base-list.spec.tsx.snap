// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseList applies sx prop 1`] = `
<ul
  class="MuiList-root css-10ly347-MuiList-root"
>
  <li
    class="MuiListItem-root css-1li41ks-MuiListItem-root"
  >
    <span>
      x
    </span>
  </li>
</ul>
`;
